storage/fcds/offsetcache.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package fcds
+
+import "sync"
+
+// offsetCache is a simple cache of offset integers
+// by shard files.
+type offsetCache struct {
```

Comment:
Would expiry be helpful to guard against memory "leak"? Is there a situation where the cache items are not removed for a long time?

Reply:
Good point. In the context of offsetCache, this is certainly a possibility, and would be a good improvement. In the wider context, removal depends on MetaStore.FreeOffset function, which is injected and cannot be trusted for any implementation. I will add basic TTL.

Reply:
I have added a basic TTL and cleanup, but I am not sure that it is actually needed while adding a bit more complexity and locking.

There is a case where offsetCashe may hold offsets for longer periods in case of no activity (uploads, retrievals or syncing). In that case expiration reduces the optimisation that offsetCache is here for, and I would not add it.

Maybe @zelig would like to share opinion on this subject also?

Reply:
I see, makes sense. I did not think that a small TTL is needed, maybe like 24h one or something similar, just to prevent cache growing forever, but I am not sure it is needed at all  as well. Do we expect to have this same cache for the whole time the app running? If yes, and if the app is supposed to be running for long periods of time (like multiple days), then it might have be good to clean it up some time. Sorry if it's a stupid question 🙈

Reply:
I have reverted the TTL feature after a brief discussion. We have concluded that if there would be a cache leak, then a much bigger problem would have cause it in this case, so the additional complexity is most likely not needed.